resilience4j-spring/src/main/java/io/github/resilience4j/retry/configure/RetryAspect.java

```diff
@@ -141,15 +143,21 @@ private Object handleSyncJoinPoint(ProceedingJoinPoint proceedingJoinPoint, io.g
 	 * @return the result object if any
 	 * @throws Throwable
 	 */
-	private Object handleAsyncJoinPoint(ProceedingJoinPoint proceedingJoinPoint, io.github.resilience4j.retry.Retry retry, String methodName) throws Throwable {
+	@SuppressWarnings("unchecked")
+	private Object handleAsyncJoinPoint(ProceedingJoinPoint proceedingJoinPoint, io.github.resilience4j.retry.Retry retry, Class<? extends RecoveryFunction> recoveryFunctionClass, String methodName) throws Throwable {
 		if (logger.isDebugEnabled()) {
 			logger.debug("async retry invocation of method {} ", methodName);
 		}
 		return io.github.resilience4j.retry.Retry.decorateCompletionStage(retry, retryExecutorService, () -> {
 			try {
 				return (CompletionStage<Object>) proceedingJoinPoint.proceed();
 			} catch (Throwable throwable) {
-				throw new CompletionException(throwable);
+				try {
```

Comment:
I think you also want to apply the recovery function when the CompletationStage completes exceptionally, or?
I guess you have to create a `CompletableFuture<T> promise = new CompletableFuture<>();`

and then 

```
decoratedCompletationStage.whenComplete((result, t) -> {
				if (result != null) {
					promise.complete(result);
				} else if (t != null) {
					promise.complete(recovery.apply(t));
				}
			});
return promise;
```

Reply:
Maybe it would be even better to add the recoveryFunction logic directly into the CircuitBreaker, Retry, Bulkhead.

Reply:
Maybe we should create a Recovery Decorator where we can implement it once and use it for Bulkhead, CircuitBreaker, RateLimiter and Retry,

Reply:
For example `Recovery.decorateCompletionStage(recoveryFunction, completationStage)`

Reply:
@Romeh I wrote this `RecoveryFunction` to maintain consistency with `RecoveryFunction` in ratpack module. But there is no problem to change it if needed.

Is this what you mean?
```java
@Recover("recoveryMethod")
public void serviceMethod(String paramter1, Integer paratemre2) {}
public void recoveryMethod(String parameter1, Integer parameter2) {}
```

Reply:
I would say just add the annotation over the recovery method itself and in that case u do not need to define method name .
public void serviceMethod(String paramter1, Integer paratemre2) {}
@Recover
public void recoveryMethod(String parameter1, Integer parameter2) {}

@RobWin what do u think ?

Reply:
What happens when someone adds `@Retry` and `@CircuitBreaker` and `@Recover` to a method.
When should the recovery function be applied? 
I think the RecoveryFunction must be attached to a concrete annotation.

Reply:
I'm also still not sure if it's an overuse of annotations.
Why not use plain old java code ;)

Reply:
imho a seperate annotation is not needed.    
The goal is to make it easier for ex-hystrix users to switch over and to simplify the code users need to write.  
So something like `@CircuitBreaker(name = "backendA", recovery = myRecoveryFunction)`
is more than enough.

Reply:
Yes it is better to attached it to the target annotation by method name and use reflection to execute the method but do not forget to validate that it has the same return type and same list of params as the retryable method

Reply:
@Romeh I think the recovery method should have one additional parameter.  
The exception that leads to the recovery function being called needs to be passed somehow, so developers can define different behaviors depending on the exception.

Reply:
one last remark , if we plan to add on result predicate check for circuit breaker into the future like the one we have on Retry right now , we need to consider that in our recovery process or not ? as there is no exception being thrown there in that case !

Reply:
Good point. I think the recovery function should be applied in that case too. Since that was a failing operation anyway. In that case, the throwable paramter could be null.
@Romeh @RobWin @123Haynes So do you think it is better to implement the `recovery` paramter as a `String` in the annotations and use reflection to call the method? And the recovery method has same signature with the service method except a throwable paratemer? Like

```java
@CircuitBreaker(name = "backendA", recovery = "recovery")
public void service(String param1, Integer param2);
public void recovery(String param1, Integer param2, Throwable throwable);
```

Reply:
@lespinsideg yeah. I think that would be great 😃

Reply:
Yes, looks good.
I still don't like this overuse of annotations and reflection magic. I would prefer a functional-style solution which can be checked by a compiler.

But we should implement it, if users of R4j request it.

Reply:
@RobWin I think aspectj is already used for all annotations, so I think the annotations can be checked by the compiler if aspectj is configured to weave things at compile time. 😃

Reply:
Right now we don't weave at compile time.

Reply:
As far as I remember we use Spring AOP and Spring AOP is proxy-based. But it's long time ago when I implemented the CircuitBreakerAspect.

See:
https://docs.spring.io/spring/docs/4.0.1.RELEASE/spring-framework-reference/htmlsingle/#aop-understanding-aop-proxies

Reply:
@RobWin I've created https://github.com/resilience4j/resilience4j/issues/391 to discuss this further.